numpy/testing/_private/utils.py

```diff
@@ -707,13 +707,18 @@ def func_assert_same_pos(x, y, func=isnan, hasval='nan'):
         """
         # Both the != True comparison here and the cast to bool_ at the end are
         # done to deal with `masked`, which cannot be compared usefully, and
-        # for which np.all yields masked.  The use of the function np.all is
-        # for back compatibility with ndarray subclasses that changed the
-        # return values of the all method.  We are not committed to supporting
-        # such subclasses, but some used to work.
+        # for which np.all yields masked. We use the all() method instead of
+        # np.all() so it still works if __array_function__ is defined but
+        # doesn't implement np.all, but use np.all() on non-numpy arrays (e.g.,
+        # booleans) that may not define an all() method. We are not committed
+        # to supporting such subclasses, but some used to work.
         x_id = func(x)
         y_id = func(y)
-        if npall(x_id == y_id) != True:
+        result = x_id == y_id
+        all_equal = (result.all()
+                     if isinstance(result, ndarray)
```

Comment:
Will this pass through subclasses? And if so, will that be an issue?

Reply:
Agree with the worry of @hameerabbasi - also, at least the comment reads a bit weird now: I don't think we have to worry about classed that define `__array_function__` but do not override `np.all` - they're new and they can override easily as part of their trials; maybe just leave it as it was?

Reply:
Yes, this passes through subclasses -- which is the point.

> I don't think we have to worry about classed that define __array_function__ but do not override np.all - they're new and they can override easily as part of their trials; maybe just leave it as it was?

I don't exactly disagree with this, but this did come up with the ndarray subclass I made for unit testing purposes. With the arrival of `__array_function__`, it will be easier to depend on built-in methods like `all()` rather than `np.all()`. I guess we offer no guarantees for subclasses that don't implement the full numpy API, but it still feels like not a great user experience.

Two other ways to do this:
1. Cast to a NumPy boolean scalar/array, which guarantees that we have an `all()` method: `np.bool_(x_id == y_id).all()`. This looks a little less hacky.
2. Give up on duck-typing for `assert_array_equal`, and instead decorate it with `array_function_dispatch` for explicit dispatching.

Reply:
In that case, if this won't cause any problems, then I agree with this design.

Reply:
Option 1 seems to cover all bases, and is shorter than what you have now.

Reply:
Hmm. Testing out masked arrays, they no longer seem to return masked arrays from `.all()` or `np.all()`:
```
In [16]: x = np.ma.array([1, 2, 3], mask=[False, True, False])

In [17]: x == x
Out[17]:
masked_array(data=[True, --, True],
             mask=[False,  True, False],
       fill_value=True)

In [18]: (x == x).all()
Out[18]: True
```

Reply:
That should cause a test failure, and if there is no test like that we should add one. Also note there are test failures with datetime64

Reply:
> That should cause a test failure, and if there is no test like that we should add one. Also note there are test failures with datetime64

I'm not sure. What would you do with a mask at all on a reduction? I mean, it makes sense to only reduce over the non-masked objects, but what is the output mask? For zero nonmasked elements do you set it to the identity of the ufunc or to `False`?